solutions/POI 14-cards.mdx

```diff
@@ -0,0 +1,92 @@
+---
+id: POI-14-cards
+source: POI
+title: Cards
+author: Dong Liu
+---
+
+**Time Complexity:** $\mathcal O(N\log N)$
+
+## Explanation
+
+We can use a segment tree to solve this problem.
+
+For each segment tree node with the range $[l, r]$, we will store if it's possible to arrange the cards $l\dots r$ so that we either
+
+- don't flip card $l$ and card $r$
+- don't flip card $l$ and flip card $r$
+- flip card $l$ and don't flip card $r$
+- flip card $l$ and card $r$
+
+We can represent these states in a 2 by 2 matrix `bool[2][2]` where the first dimension stores whether we flip card $l$ and the second dimension stores whether we flip card $r$.
+
+To calculate the states for a specific segment tree node, we loop through all possible states of our left child node $(a, b)$ and all possible states of our right child node $(c, d)$. If the transition from $b$ to $c$ works, then $(a, d)$ works since we're only considering endpoints.
+
+To perform a swap operation with index $i$ and $j$, we simply swap the cards at $i$ and $j$. We can just recalculate all nodes in the segment tree that are related to either $i$ or $j$. Since there is around $O(\log N)$ ranges in the segment tree that contain $i$ or $j$, this whole swap operation takes $O(\log N)$.
```

Comment:
```suggestion
To perform a swap operation with index $i$ and $j$, we simply swap the cards at $i$ and $j$. We can just recalculate all nodes in the segment tree that are related to either $i$ or $j$. Since there are around $O(\log N)$ ranges in the segment tree that contain $i$ or $j$, this whole swap operation takes $O(\log N)$.
```